content/blog/2019-11-21-transmeta.md

```diff
@@ -0,0 +1,126 @@
++++
+title = "The Transmeta Code Morphing Software"
+
+[[extra.authors]]
+name = "Ryan Doenges"
+link = "http://ryandoeng.es/"
+
+extra.bio = """
+  [Ryan Doenges](http://ryandoeng.es/) is a third year PhD student studying
+  programming languages and networking.
+"""
++++
+
+Today we are reading the 2003 paper on Transmeta CMS (Code Morphing
+Software). The CMS layer ran x86 programs on the Transmeta Corporation's Crusoe
+microprocessor, which had an internal architecture that was much simpler than an
+x86. While much of the terminology in the paper is non-standard, I hope it was
+clear that CMS is a just-in-time (JIT) compiler for x86 targeting Crusoe's
+internal instruction set architecture (ISA).
+
+## Why a JIT?
+How and when the computer industry settled on (for?) the x86 instruction set
+architecture I do not know, but we can surmise from the engineering effort
+expended by Transmeta that it happened before 2003. At the time, new general
+purpose processors needed to expose an x86 interface to programmers.
+
+Faced with this task, Transmeta engineers could have gone the obvious route and
+built an x86 clone in hardware. The paper argues for the internal ISA and CMS
+technique as follows.
+
+> This approach allows a simple, compact, low-power microprocessor
+> implementation, with the freedom to modify the internal ISA between
+> generations, while supporting the broad range of legacy x86 software
+> available.
+
+To flip all these adjectives, the Crusoe designers recognized that a direct
+x86 implementation like Intel's was complicated, sprawling, and high-power. They
+understood that Intel's infamous commitment to backwards compatibility was good
+for business but made it difficult to modify hardware when it might improve
+maintainability, space usage, or power efficiency. Hiding their actual
+architecture behind an x86 abstraction solved this problem.
+
+## The internal ISA
+
+Crusoe's internal ISA is a VLIW (very long instruction word) ISA with 64
+general-purpose registers and 32 floating-point registers, which is more than
+the x86. In a VLIW instruction set like Crusoe's, each instruction is really
+several smaller instructions which are issued in parallel. In the terminology of
+the paper the large instructions are "molecules" composed of 2 or 4 "atoms".
+The internal ISA avoids handling pipeline stalls--instead it expects the CMS
```

Comment:
I know this is silly, but I think `---` is better for em dashes. (I wish Zola supported actually converting these to —. Maybe someday.)